By default turn on all `DeprecationWarning` at SymPy level.

Carreau opened the pull request:
Apply that only to dev version (let's not be too aggressive),
and turn them into error on Travis (if continuous integration).
Otherwise you'll miss API's updates.

---

ping @asmeurer 

```
$ isympy
REPO/sympy/core/function.py:107: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  evalargspec = inspect.getargspec(cls.eval)
REPO/sympy/core/function.py:107: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  evalargspec = inspect.getargspec(cls.eval)
REPO/sympy/interactive/session.py:475: DeprecationWarning: mainloop `display_banner` argument is deprecated since IPython 5.0. Call `show_banner()` if needed.
  mainloop(message)

In [1]:
```

From a recent [disagreement ](https://github.com/ipython/ipython/pull/9631) we had on how to handle deprecations. 
Who knows the test might turn up more deprecated features.

Carreau committed:
By default turn on all `DeprecationWarning` at SymPy level.

Apply that only to dev version (let's not be too aggressive),
and turn them into error on Travis (if continuous integration).
Otherwise you'll miss API's updates.

asmeurer commented on sympy/__init__.py:
```diff
@@ -22,6 +22,21 @@
+        if travis:
+            action = 'error'
+        warnings.filterwarnings(action,   '.*',                      DeprecationWarning, module='sympy.*')
+        # let's not fail the test on that yet. 
```
Can you explain `inspect.signature` was added in Python 3.3, so there's no way to avoid the warning in Python 2.7.

Carreau replied:
I'm confused about this statement. Avoid which warning ? on 2.7 getargspec is not deprecated so you wont' get the warning.

asmeurer replied:
I mean there's no way to avoid the warning while supporting Python 2.7. Although I suppose we could do different things in 2 and 3.

Carreau replied:
I can swap for an 'ignore' and back to default if `sympy.__version__ > (1,5)` so that you can revisit this this deprecation later.

asmeurer replied:
If you don't want to address this now, open an issue. Don't put a sleeper in the code.

Carreau replied:
Opened. https://github.com/sympy/sympy/issues/11255

I was first trying to get the test-suite to pass.

Carreau committed:
Fix use of fraction.gcd [deprecated in 3.5]

replace by `math.gcd`. Note that math.gcd only deals with positive
integers, but that's what Egyptian fraction deals with anyway.

Also remove trailing space.

Carreau commented:
@asmeurer needs anything else from me ?

asmeurer commented on sympy/__init__.py:
```diff
@@ -22,6 +22,21 @@
+        import os
+        import warnings
+        action ='default'
+        travis = (os.environ.get('TRAVIS','') == 'true')
```
I think it would be better to always fail when the tests are running (or always). Things that are only reproducible in Travis are annoying.

Carreau replied:
Any simple way to know we are running the tests ? setting `default` by default seem useless as test will never fails, and setting `error` seem annoying for local user where Sympy would not even import.

asmeurer replied:
It would need to be set in the test runner (sympy/utilities/runtest.py), I guess in the `_test` function.

asmeurer replied:
If you search the file for "warnings" there is already some code there to enable SymPyDeprecationWarning.

Carreau committed:
Use signature on python3 since getargspec is deprecated since 30.

And getfullargspec is deprecated since 3.5.

Carreau commented:
Sidenote while I'm working on that: 

I know the Sympy project and that if I make a PR I will be welcomed, but I also strongly believe that having tens of page of pull-requests can look unfriendly to new contributors. That's basically from experience, when I came across an new project, if I see more than a dozen of pull requests I tend to think "they will never bother with mine ;  they  probably don't even look at the list". If I may suggest, I would tag the inactive PR with a specific tag like `happy to resurrect`, `on hold for inactivity` and actually close them or something similar. 

Every project have it's own ecosystem and social rules, just giving a data point.

asmeurer commented:
Yeah, it's a problem.

Carreau committed:
Cleanup implementation and eval args lazily.

Carreau committed:
Turn inspect.getargspec, DeprecationWarnings back into error.

We should not call it anymore.
By default show the Sympy DeprecationWarnigns to users, don't raise
though.

Carreau committed:
Show the warnings interactively, don't crash in user hands.

Carreau commented:
I should have taken care of the `inspect.signature` case I well, I took the opportunity to make the computation of number of argument lazy has it was not always used.

Carreau commented:
@asmeurer can you relaunch the tests for all the 3.5 instances, Travis seem to have broken their 3.5.2 installs close to the time I pushed.

Carreau commented:
Travis seem to be happy now !

Carreau commented:
> So do you think we should change https://github.com/sympy/sympy/blob/2a1ab90760f20e3b436eea428ceadf32e9f9a1e7/sympy/utilities/runtests.py#L495 and https://github.com/sympy/sympy/blob/2a1ab90760f20e3b436eea428ceadf32e9f9a1e7/sympy/utilities/runtests.py#L680 from SymPyDeprecationWarning to DeprecationWarning?

Not sure, that's not exactly the same. 

1 ) The line you refer to  change `SymPyDeprecationWarnings` that sympy **emits** to be errors regardless of where they are triggered. These might be triggered during the Sage Integration test for example, in which case changing this line would likely turns the error off. 

Typically if [`sage.euler_gamma`](https://github.com/sympy/sympy/blob/master/sympy/external/tests/test_sage.py#L120) were to use a Deprecated SymPy warning, SymPyDeprecationWarnings as error would catch it, not this PR.

So they basically force your dependents to update. 

2 ) This pull requests change DeprecationWarning that SymPy **triggers** into error regardless of where they come from. 

So they basically force you to update if your dependencies update their API.

I think both approaches are complements of each others, and that 1) should definitively have a generic Python API to filter for. 

Does that makes sens ?

asmeurer commented:
Those lines are specifically to make warnings errors when the tests run.

Carreau commented:
> Those lines are specifically to make warnings errors when the tests run.

I know. I'm just telling you that `warnings.simplefilter("error", SymPyDeprecationWarning)`  and `warnings.filterwarnings('default',   '.*',   DeprecationWarning, module='sympy.*')` turn different kind of warnings in error. 

# Type A warning:


```
# file sage/something.py
def function_that_uses_deprected_sympy_api(*args):
      sympy.you_don_t_know_it_but_i_m_deprecated() # raise SymPyDeprecationWarning
```

 - Will be turned into an error by `warnings.simplefilter("error", SymPyDeprecationWarning)`
 - will not be turned into an error by `warnings.filterwarnings('default',   '.*',   DeprecationWarning, module='sympy.*')`

Example `Particle.get_mass` might be used by AstroPy

# Type B warnings

```
# file sympy/something.py

def sympy_function_that_uses_deprcated_api(*args):
           you_don_t_know_it_but_i_m_also_deprecated() # likely raise a DeprecationWarning not a SympyOne. 
```

 - Will **not** be turned into an error by `warnings.simplefilter("error", SymPyDeprecationWarning)`
 - Will be turned into an error by `warnings.filterwarnings('default',   '.*',   DeprecationWarning, module='sympy.*')`


Example: sympy use `inspect.getargspec`.


# Recap:

The intersection exists when SymPy uses it's own deprecated API. You are may be testing for Type B as you are installing matplotlib, numpy, scipy, sagemath... etc and calling it, but maybe you are not. 
I know that matplotlib does not use sympy, I don't know for sage. 
If none of these packages directly use SymPy, Then it should not matter to remove these lines.  I just don't know if the actual set is disjoint, or Type B subset of Type A. In the general case they are both useful. 

| Producer  \Consumer | Sympy        | Other  |
|---------------------|--------------|--------|
| Sympy               | Type A and B | Type A |
| Other               | Type B       | N/A    |
|                     |              |        |

To counter my own argument {Type B} - {Type A} should only catch bugs on non-sympy code, you you should be safe reactivating `warnings.simplefilter("error", SymPyDeprecationWarning)`.

We also might just be misunderstanding each other.

asmeurer commented:
My point is that when the tests are being run, all deprecation warnings (SymPy or otherwise) should be errors. Otherwise issues of the type that you fixed in this PR will re-appear (and people ignore error messages printed during the tests if they don't cause the tests to fail).

Carreau committed:
Restore Deprecation warnings are error during test.

Carreau commented:
I yes ! I see you are right.  I forgot about that when I turned back the warning to `default` in https://github.com/sympy/sympy/pull/11249/commits/212c363dd14521e7c2b10f550c047282880c4bbe . 

I've readded `warnings.filterwarnings('error', '.*', DeprecationWarning, module='sympy.*')` in addition to `warnings.simplefilter("error", SymPyDeprecationWarning)`. 

I feel dumb now.

asmeurer commented:
You should do it for the doctests too (the other line I mentioned).

Carreau committed:
make DeprecationWarnings errors in doctests as well

Carreau committed:
Fix getargspec usage in Theano related code.

Carreau committed:
Fix one more deprecated usage of getargspec.

Carreau commented:
Travis is Happy, again.

asmeurer commented:
OK, me too :)

asmeurer commented:
Just realised that all the compatibility stuff should have been put in the compatibility file.

Carreau commented:
Oh, sorry about that.

We can do that during sprints.

On Tue, Jul 12, 2016 at 11:22 AM, Aaron Meurer <notifications@github.com>
wrote:

> Just realised that all the compatibility stuff should have been put in the
> compatibility file.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/11249#issuecomment-232099812>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAUez-W-NF7G7WBsdN9r3oHwflh298rbks5qU79agaJpZM4I41cG>
> .
>

asmeurer commented:
Opened https://github.com/sympy/sympy/issues/11412 so this gets fixed.

Carreau commented:
@asmeurer https://pypi.python.org/pypi/warn as a prototype, woudl you are to have a look . 
(warning code is still ugly) 

Feedback welcome, if I get it in a more useful state I might ask for a few changes to Python-dev.
